runtime/src/bank.rs

```diff
@@ -3455,25 +3463,80 @@ impl Bank {
         consumed_budget.saturating_sub(budget_recovery_delta)
     }
 
+    pub fn secp256k1_program_enabled(&self) -> bool {
+        self.feature_set
+            .is_active(&feature_set::secp256k1_program_enabled::id())
+    }
+
     // This is called from snapshot restore AND for each epoch boundary
     // The entire code path herein must be idempotent
     fn apply_feature_activations(&mut self, init_finish_or_warp: bool, initiate_callback: bool) {
-        self.ensure_builtins(init_finish_or_warp);
+        let new_feature_activations = self.compute_active_feature_set(!init_finish_or_warp);
+        self.ensure_builtins(init_finish_or_warp, &new_feature_activations);
         self.reinvoke_entered_epoch_callback(initiate_callback);
         self.recheck_cross_program_support();
         self.recheck_compute_budget();
         self.reconfigure_token2_native_mint();
         self.ensure_no_storage_rewards_pool();
     }
 
-    fn ensure_builtins(&mut self, init_or_warp: bool) {
-        for (program, start_epoch) in get_builtins(self.cluster_type()) {
+    // Compute the active feature set based on the current bank state, and return the set of newly activated features
+    fn compute_active_feature_set(&mut self, allow_new_activations: bool) -> HashSet<Pubkey> {
+        let mut active = self.feature_set.active.clone();
+        let mut inactive = HashSet::new();
+        let mut newly_activated = HashSet::new();
+        let slot = self.slot();
+
+        for feature_id in &self.feature_set.inactive {
+            if let Some(mut account) = self.get_account(feature_id) {
+                if let Some(mut feature) = Feature::from_account(&account) {
+                    match feature.activated_at {
+                        None => {
+                            if allow_new_activations {
+                                // Feature has been requested, activate it now
+                                feature.activated_at = Some(slot);
+                                if feature.to_account(&mut account).is_some() {
+                                    self.store_account(feature_id, &account);
+                                }
+                                newly_activated.insert(*feature_id);
+                                active.insert(*feature_id);
+                                info!("Feature {} activated at slot {}", feature_id, slot);
+                                continue;
```

Comment:
nits: I think these two `continue`s' control flow is a bit hard to follow. At the same time, I feel the urge to skip the general catch-all `inactive.insert(...)` for these cases in some way. So, I'd like this to be something like this:

```rust
diff --git a/runtime/src/bank.rs b/runtime/src/bank.rs
index 078d2402e4..daab6dd347 100644
--- a/runtime/src/bank.rs
+++ b/runtime/src/bank.rs
@@ -3484,6 +3484,7 @@ impl Bank {
         let slot = self.slot();
 
         for feature_id in &self.feature_set.inactive {
+            let mut activated = false;
             if let Some(mut account) = self.get_account(feature_id) {
                 if let Some(mut feature) = Feature::from_account(&account) {
                     match feature.activated_at {
@@ -3497,20 +3498,22 @@ impl Bank {
                                 newly_activated.insert(*feature_id);
                                 active.insert(*feature_id);
                                 info!("Feature {} activated at slot {}", feature_id, slot);
-                                continue;
+                                activated = true;
                             }
                         }
                         Some(activation_slot) => {
                             if slot >= activation_slot {
                                 // Feature is already active
                                 active.insert(*feature_id);
-                                continue;
+                                activated = true;
                             }
                         }
                     }
                 }
             }
-            inactive.insert(*feature_id);
+            if !activated {
+                inactive.insert(*feature_id);
+            }
         }
 
         self.feature_set = Arc::new(FeatureSet { active, inactive });
```

Reply:
I further edited my diff; specifically moved the `active` handling as well. This way, the dichotomy of `active` and `inactive` is clear in my opinion. :)

Reply:
Thanks, I applied your suggestion at https://github.com/solana-labs/solana/pull/12510